crates/sui-framework/sources/collectible/kiosk.move

```diff
@@ -0,0 +1,564 @@
+/// be used to implement application-specific transfer rules.
+///
+module sui::kiosk {
+    use std::option::{Self, Option};
+    use sui::object::{Self, UID, ID};
+    use sui::dynamic_object_field as dof;
+    use sui::dynamic_field as df;
+    use sui::transfer_policy::{Self, TransferRequest};
+    use sui::tx_context::{TxContext, sender};
+    use sui::balance::{Self, Balance};
+    use sui::coin::{Self, Coin};
+    use sui::sui::SUI;
+    use sui::event;
+
+    // Collectible is a special case to avoid storing `Publisher`.
+    friend sui::collectible;
+
+    /// Trying to withdraw profits and sender is not owner.
+    const ENotOwner: u64 = 0;
+    /// Coin paid does not match the offer price.
+    const EIncorrectAmount: u64 = 1;
+    /// Trying to withdraw higher amount than stored.
+    const ENotEnough: u64 = 2;
+    /// Trying to close a Kiosk and it has items in it.
+    const ENotEmpty: u64 = 3;
+    /// Attempt to take an item that has a `PurchaseCap` issued.
+    const EListedExclusively: u64 = 4;
+    /// `PurchaseCap` does not match the `Kiosk`.
+    const EWrongKiosk: u64 = 5;
+    /// Tryng to exclusively list an already listed item.
+    const EAlreadyListed: u64 = 6;
+    /// Trying to call `uid_mut` when extensions disabled
+    const EExtensionsDisabled: u64 = 7;
+
+    /// An object which allows selling collectibles within "kiosk" ecosystem.
+    /// By default gives the functionality to list an item openly - for anyone
+    /// to purchase proviging the guarantees for creators that every transfer
+    /// needs to be approved via the `TransferPolicy`.
+    struct Kiosk has key, store {
+        id: UID,
+        /// Balance of the Kiosk - all profits from sales go here.
+        profits: Balance<SUI>,
+        /// Always point to `sender` of the transaction.
+        /// Can be changed by calling `set_owner` with Cap.
+        owner: address,
+        /// Number of items stored in a Kiosk. Used to allow unpacking
+        /// an empty Kiosk if it was wrapped or has a single owner.
+        item_count: u32,
+        /// Whether to open the UID to public. Set to `true` by default
+        /// but the owner can switch the state if necessary.
+        allow_extensions: bool
+    }
+
+    /// A Capability granting the bearer a right to `place` and `take` items
+    /// from the `Kiosk` as well as to `list` them and `list_with_purchase_cap`.
+    struct KioskOwnerCap has key, store {
+        id: UID,
+        for: ID
+    }
+
+    /// A capability which locks an item and gives a permission to
+    /// purchase it from a `Kiosk` for any price no less than `min_price`.
+    ///
+    /// Allows exclusive listing: only bearer of the `PurchaseCap` can
+    /// purchase the asset. However, the capablity should be used
+    /// carefully as losing it would lock the asset in the `Kiosk`.
+    ///
+    /// The main application for the `PurchaseCap` is building extensions
+    /// on top of the `Kiosk`.
```

Comment:
This is really helpful, thank you! (It's possible I missed this on first reading as well... not sure)

Reply:
This was added upon feedback from us (originbyte) as well as some other patterns ^.^

Reply:
Your input won't be left unmentioned!